File: chap19_retina/direction_detection.m
% Code to simulate the direction detection model
clear all
close all
clc
set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 2.0, ...
   'defaultlinelinewidth', 2.0, ...
   'defaultpatchlinewidth', 0.7);

%parameters
Dx = 0.5;
Dt = 0.2;
c = 0.2;

t=[-20:0.1:20];
cval = [0.2,-0.2];
for j = 1:2
    c = cval(j);
R = S(-c*t).*S(Dx-c*t+c*Dt)-S(Dx-c*t).*S(-c*t+c*Dt);
figure(1) 
plot(t,R)
hold on
xlabel('t')
ylabel('R(t)')
end
legend('boxoff')
legend('c=0.2','c=-0.2','location','northwest')
box off

hold off


function out =S(x)

out = 10*(1+exp(-(x-1).^2) + 0.8*exp(-(x+0.2).^2));
end

File: chap19_retina/direction_detection.py

#  ---------------------------
#  The Reichardt direction detection model 
# 
#  Used to generate the image in Fig. 20 of Chapter 19 of
#  Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#  Written by James Keener and James Sneyd
#  ---------------------------

import numpy as np
import matplotlib.pyplot as plt

# Parameters
Dx = 0.5
Dt = 0.2
c_values = [0.2, -0.2]

# Time vector
t = np.arange(-20, 20, 0.1)

# Function S(x)
def S(x):
    return 10 * (1 + np.exp(-(x - 1) ** 2) + 0.8 * np.exp(-(x + 0.2) ** 2))

# Plotting
plt.figure(figsize=(10, 5))
for c in c_values:
    R = S(-c * t) * S(Dx - c * t + c * Dt) - S(Dx - c * t) * S(-c * t + c * Dt)
    plt.plot(t, R, label=f'c={c}')

plt.xlabel('t')
plt.ylabel('R(t)')
plt.legend()

